### ISSUE TITLE
Strange .splice behaviour

### ISSUE THREAD

**Issue by bykof** [solkimicreb/react-easy-state#55]
Hi guys, 

my store looks like this: 
```
const PostStore = store(
    {
        posts: [],
    }
);
```

this is how I render them: 
```
{
    PostStore.posts.map(
        (post) => <PostComponent key={post._id} post={post}/>
    )
}
```

I have a socket which is connected to the backend and everytime a post is saved it does:
_this is the only place where posts changed_ 
```
PostStore.posts.splice(0, 0, post);
```

But the problem is, that the react component is *appending* the posts.
So if there were like 2 posts before, it will render 5 because one was added and the other on top too.

Can someone explain me why react does not destruct the old posts?

**Comment by bykof** [comment]
It looks like it calls the render method of ListComponent (where the posts are rendered) for every post in `PostStore.posts` when I do `splice`

**Comment by bykof** [comment]
When I am using `.push(post)` it works and it renders only one time.

**Comment by solkimicreb** [comment]
Hi!

Are you mutating your data from a websocket event listener?

**Comment by bykof** [comment]
Yes, from socket.io. 
Why did you close the ticket? :(

**Comment by bykof** [comment]
This is my socket:
```
import io from "socket.io-client";

import PostStore from '../stores/postStore';


class PostSocket {
    constructor() {
        this.socket = null;
    }

    get connected() {
        return !!this.socket;
    }

    connect() {
        this.socket = io('http://localhost:5000');
        this.socket.on('post.saved', this.postSaved);
    }

    postSaved(post) {
        const postIndex = PostStore.posts.findIndex((postTemp) => postTemp._id === post._id);
        if (postIndex > -1) {
            PostStore.posts[postIndex] = post;
        } else {
            PostStore.posts.push(post);
        }
    }
}
const postSocket = new PostSocket();
export default postSocket;
```

**Comment by solkimicreb** [comment]
Sorry, I misclicked 🤦‍♂️ .

This was an oversight from me. Multiple rendering is caused, because I forgot to add websocket events to the batched task sources (behind the scenes splice is doing tons of mutations). The appending issue seems to be something different though. I am pretty sure it is not caused by Easy State.

I will add websocket to the batched sources tomorrow. In the meantime you can use manual batching in this case, if you would like to. Check this [docs section](Semmelweis utca 19).

**Comment by bykof** [comment]
I am on version 6.0.2 but there is no batch exported in the react-easy-state o.O
@solkimicreb

**Comment by solkimicreb** [comment]
Super strange 😕 All tests passed for all builds, but I guess I messed up something with the publishing.

I don't have my laptop with me now, but I will check this ASAP. Sorry for the mess.

**Comment by bykof** [comment]
No problem, the most important thing is that we found an error :)

**Comment by Penspinner** [comment]
Weird. I just tried importing `batch` and it works fine.

https://codesandbox.io/s/jvw4yp004v

**Comment by bykof** [comment]
DUDE! While I am updated to 6.0.2 I recognized a *fucking really big problem*!
You are overwriting `Promise.prototype.catch = batchCallbacks(Promise.prototype.catch);` in [scheduler.js](https://github.com/solkimicreb/react-easy-state/blob/77424494c8c48a0f2d2e51ceaa97765812494823/src/scheduler.js#L71)
This blows up my whole application!

I am tried just the simple example of: 
```
function a(){
    return new Promise(
        (resolve, reject) => {
            reject(new Error('a new error!'));
        }
    )
}

a().then().catch((error) => {console.log('my-error', error)});
```

It prints me the function as a string: 
```
ƒ (error) {
  console.log(error);
}
```

**Comment by bykof** [comment]
With your prototype patches I can't do any promises! 
:D Let's try to fix this

**Comment by bykof** [comment]
Until you fix the problem. 
Put this at the top of your html: 
```
<script type="text/javascript">
    var oldPromiseThen = Promise.prototype.then;
    var oldPromiseCatch = Promise.prototype.catch;
</script>
```

and then in your index.js
```
/* global oldPromiseThen oldPromiseCatch */
Promise.prototype.then = oldPromiseThen;
Promise.prototype.catch = oldPromiseCatch;
```

**Comment by solkimicreb** [comment]
Fixed the critical issue in v6.0.3 (I kept the auto batching). I will get back to websocket batching tomorrow. Thanks for the help @bykof and @Penspinner

**Comment by bykof** [comment]
Alright it works! 
I can import batching, my promises are working again and everything is fine. Thank you for the really quick fix hacktoberfest is running! :) 
Have a great day!

**Comment by solkimicreb** [comment]
@bykof I just published v6.0.4, which has auto batching for all websocket handlers. Please try it when you have some time. Does importing `batch` work for you btw? I can't reproduce an issue with it.

Side note: If you are working on something not so big and open source with websockets and easy-state, I would love to include it as an example in the repo readme.

**Comment by bykof** [comment]
@solkimicreb sorry for the late answer.
I updated to v6.0.4 and it worked *without* batch! 

> Does importing batch work for you btw?
- Yes it worked for me before and after v.6.0.4

> Side note: If you are working on something not so big and open source with websockets and easy-state, I would love to include it as an example in the repo readme.
- No sorry, my project is not open source. If it were than I would include it :) Maybe in another future project, i keep that in mind.

Sincerely, Michael
